juju/juju
# [JUJU-1334] Reduce http client instances

This reuses the HTTP client for Charmhub API requests so we can take advantage of HTTP connection reuse. Because deploys do about 5-6 requests to the Charmhub API, and bundles of course do numCharms times that, it makes a pretty big difference, especially from APAC, which is far from the Charmhub API server: subsequent requests to the API are about 350ms instead of ~1s each.

This cuts the time for `juju deploy charmed-kubernetes` down from about 1 minute to approximately 30s in my tests.

`HTTP_PROXY` settings should still work fine, as this continues to use the `juju/http` package, which (by default) checks the proxy environment variables every request (unlike the default `net/http` client, which only does it once).

## QA Steps

```
$ juju bootstrap lxd
$ time juju deploy charmed-kubernetes
...
real	0m30.502s
```

To see the actual request times, I used this diff to add some simple request logging:

```diff
$ git diff
diff --git a/charmhub/http.go b/charmhub/http.go
index 0e62e36ff7..bd8ae639e8 100644
--- a/charmhub/http.go
+++ b/charmhub/http.go
@@ -127,7 +127,10 @@ func NewAPIRequester(transport Transport, logger Logger) *APIRequester {
 // Do performs the *http.Request and returns a *http.Response or an error
 // if it fails to construct the transport.
 func (t *APIRequester) Do(req *http.Request) (*http.Response, error) {
+       start := time.Now()
        resp, err := t.transport.Do(req)
+       elapsed := time.Since(start)
+       t.logger.Errorf("TODO APIRequester.Do %s %s took %s", req.Method, req.URL.String(), elapsed)
        if err != nil {
                return nil, errors.Trace(err)
        }
```

And saw that the first request was ~1s and subsequent requests were ~350ms.
